### PR TITLE
Sanitation of secrets from returner output

### PR DESCRIPTION
**Pull request by kmshultz**
This feature allows a user to prevent secrets in pillar from winding stored in any sensitive datastores via returners.

Here's a partial `minion_config` which uses the feature:

``` yaml
# specify the pillar namespaces which contain the secrets you'd like redacted.
# this can be a string (for a single namespace) or a list.
returner_sanitize.secrets_pillars:
  - master:pillar_roots

# specify any "secrets" expected in your secrets_pillars above which are
# benign and need not be redacted. e.g. Perhaps your pillar contains
# some common words, such as 'root', which could make a mess
# of many job outputs if redacted
returner_sanitize.secrets_whitelist:
  - /srv/spm/pillar

# all returners' `returner` functions will be sanitized, except 
# those returners in this list
returner_sanitize.returner_whitelist:
  - local_cache
```

The only config option _required_ to activate the feature is `returner_sanitize.secrets_pillars`.

I tested locally using a simple script, `sanitize-returner.py`:

``` python
from salt.cli.caller import BaseCaller

runner = BaseCaller(salt.config.minion_config('/Users/kent/minion_config'))
runner.opts['fun'] = 'pillar.item'
runner.opts['arg'] = [sys.argv[1]]
runner.opts['local'] = True
runner.opts['metadata'] = None
runner.opts['return'] = 'local'
runner.run()
```

And here's an example run (which uses the `local` returner, as you can see):

```
$ ~/sanitize-returner.py master:pillar_roots
{u'fun_args': [u'master:pillar_roots'], u'jid': u'20151028151656874983', u'return': {u'master:pillar_roots': {u'base': [u'REDACTED', u'/srv/spm/pillar', u'REDACTED']}}, u'retcode': 0, u'success': True, u'fun': u'pillar.item', u'id': u''}
local:
    ----------
    master:pillar_roots:
        ----------
        base:
            - /srv/pillar
            - /srv/spm/pillar
            - /srv/pillar/foo
```

I've played around with the script and config a bit to test the code's features as thoroughly as I can, but I haven't written any tests for the feature yet.

I haven't filed an Issue for this; I can do that if you like. Please let me know what you all think!

### CODE REVIEW

File: salt/loader.py [hunk 1]
```diff
@@ -1574,3 +1646,35 @@ def wrapper(*args, **kwargs):
...
+                msg = ('While sanitizing a job\'s output, its \'ret\''
+                       ' JSON was mangled (i.e. it\'s no longer parseable'
+                       ' JSON). Hence, the job result has *NOT* been published'
+                       ' to the configured returners. The mangled \'ret\''
+                       ' string in question is:\n\n{0}\n\nPlease review the'
+                       ' \'returner_sanitize.secrets_whitelist\' config'
+                       ' option; You can probably find some benign secret to'
+                       ' add there to fix the bad JSON.'.format(json_string))
+                log.error(msg)
+                raise SaltException(message=msg)
```
Comment (kmshultz) [inline salt/loader.py]: Should we make a custom Exception type for this?

**Comment by kmshultz** [comment]
The LazyLoader build failures are due to https://github.com/saltstack/salt/pull/28389.

**Comment by cachedout** [comment]
I've been looking at this today and so far it looks great! I would definietely like to see us get some test coverage in ASAP for this, so if you'd like to open another issue on that, it would be great. I'm now going to review and hopefully merge #28389 and then we'll re-run the tests on this one. Thanks very much for your good work here. I think this will be an excellent addition to Salt!

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
@jacksontj This might be of interest to you as well.

**Comment by kmshultz** [comment]
Thanks @cachedout, I'll open up an Issue soon and take a look at testing.

**Comment by thatch45** [comment]
I really like this, but given that it is security centric in nature we should have some tests in place for it before it is merged in, as well as some docs.
The tests should likely be in this PR but we can get the docs in later on.

**Comment by cachedout** [comment]
I added the `Pending Changes` label to indicate that we'll need to have test coverage written before we can merge this.

**Comment by cachedout** [comment]
@kmshultz Did you see the comments from @thatch45 yesterday? Just pinging you to make sure.

**Comment by kmshultz** [comment]
@cachedout yes, I saw the comments. I haven't had time to look at this really since I submitted the PR, but I'm hoping to write some tests soon (maybe in the next week). I can also write the documentation if/when the time comes.

Just wanted to run this by you guys first - glad you like it!

**Comment by cachedout** [comment]
@kmshultz No problem. Just making sure you saw them. Thanks!

**Comment by jacksontj** [comment]
This is an interesting idea, I do have a couple suggestions though:

(1) Configuring whether an item is secret in the master config seems a bit clunky for the general case-- since usually people are mutating the pillar without mutating the master config. In larger scale environments its unreasonable to require a master restart to pick up new redacted secrets. Maybe this should be chained with a renderer such that when the items are rendered the master can maintain some other namespace of all the secrets it rendered. AN example would be some magic within the gpg renderer to keep track of all the things it decrypted, and add an item to the pillar something like ``__gpg_decrypted_things``. Something like that would make it easier to determine what is encrypted (since it would be flagged somehow) as well as more likely to be used (since it won't require master config changes).

(2) It might be better to require people to set the returner to your returner instead of adding the decorator magically in the loader. The loader already has a decent amount of magic, and its quite easy to wrap returners in other ones (example; https://github.com/saltstack/salt/blob/develop/salt/returners/multi_returner.py). This would leave the loader cleaner-- and it would also mean you can develop and iterate on the feature in module space.

**Comment by cachedout** [comment]
@kmshultz Just checking in with you on this one. No hurry of course, but please let me know where we are on it. Thanks!

**Comment by kmshultz** [comment]
@jacksontj Thanks for the feedback. In my use case, I'm trying to redact secrets found in an external pillar, so renderers don't come into play there. Also, the external pillar's whole purpose is that it's the repository of all secrets in the environment, so any new secrets will be placed there as well. Hence, the config exposed by this PR should rarely change (i.e. rarely require a master restart) for me. Of course I realize not everyone's environment works this way.

So, while the renderer idea can work for pillars rendered from SLS files in pillar_root, it's not comprehensive enough to cover all of pillar (e.g. external pillars), right?

I can't see a better way for saltmaster to reliably anticipate what secrets need redacting other than by having a list of pillar namespaces passed to it. The trick is for users to be able to refresh that list without needing to restart saltmaster, as you've pointed out.

Also, it'd be nice to redact older secrets as well (up to some point, whether it's by date, number of rotations ago, etc). This PR doesn't provide that.

Thoughts?

**Comment by cachedout** [comment]
@jacksontj Just a ping to ensure you saw the above reply from @kmshultz. Looking forward to your reply. Thanks!

**Comment by cachedout** [comment]
@jacksontj Apologies for the repeated pings. Just hoping for some feedback for @kmshultz when you get a minute. Many thanks.

**Comment by cachedout** [comment]
Hi @kmshultz . Apologies for the delay. Doesn't look like we've heard back from @jacksontj here.

How would you like to proceed? Do you feel comfortable for this approach for the time being? I have no problem with getting this in as it sits but would like to check in with you first. Thanks.

**Comment by jacksontj** [comment]
Apologies for the long delay, its been a busy few weeks. In addition to not having to restart the master this feels incomplete as mentioned before. My main concern is that this requires to you scope all of your secrets to a set of SLS files-- and all things in those SLSs are redacted. IMO ideally we'd be able to mark each item in the pillar as "redacted" since you wanting it redacted is usually not tied to being within a specific SLS but rather the actual value (passwords etc.).

An idea for doing so would be to effectively "tag" all redacted items in the pillar. I don't have time today to hack out a PoC, but I can explain the idea.

Right now the pillar system gets a dict from some source (files, ext_pillar, etc.) and ships it to the minion upon request. The minion then passes that dict around to other things. And in your case above you overrode the returner so it would iterate over the pillar looking for particular values. We'll need to keep that-- my idea would be to change how we determine what should be redacted. If we created some wrapper object (similar to how kwargs are packed into a dict with a __ tagged item in the publish world) we can pass it around, and then re-serialize it as a redacted item on the remote end.

So, a simplistic example of how the client could work:
```
class Redacted(object):
    pass

def redacted(item):
    class RedactedItem(item.__class__, Redacted):
        pass
    return RedactedItem(item)

things = (
    1,
    {'a': 'b'},
    [1,2,3],
    'a'
)

for item in things:
    r = redacted(item)
    print item, r
    print isinstance(r, type(item)), isinstance(r, Redacted)
    print
```

This way the item still acts as the pillar item, but we can determine if it was intended to be redacted. This would mean that the pillar would have to include some other markup such that it could be serialized before being served to a minion. 

An alternative to the class based markup would be to have the pillar modules return a list of redacted keys-- so then they can determine their own redacted secrets (basically change the return type of ext_pillar to return a dict and a list-- pillar and redacted-keys).

Something to solve the "which things should be redacted" in a more dynamic way would go a long way in adoption for this feature.

**Comment by cachedout** [comment]
@kmshultz Did you see the above comment?

**Comment by thatch45** [comment]
Due to the inactivity in the PR I am going to close it out. Please feel free to reopen it if needed

**Pull request closed without merging**